django/db/models/query.py

```diff
@@ -594,18 +634,79 @@ def raw(self, raw_query, params=None, translations=None, using=None):
                 params=params, translations=translations,
                 using=using)
 
+    def _setup_values_query(self):
+        """
+        Constructs the field_names list that the values query will be
+        retrieving.
+        """
+        query = self.query
+        query.select_related = False
+        if query.group_by is True:
+            query.add_fields([f.attname for f in self.model._meta.concrete_fields], False)
+            query.set_group_by()
+        query.clear_deferred_loading()
+        query.clear_select_fields()
+
+        if self._fields:
+            extra_names = []
+            annotation_names = []
+            if not query._extra and not query._annotations:
+                # Shortcut - if there are no extra or annotations, then
+                # the values() clause must be just field names.
+                field_names = list(self._fields)
+            else:
+                field_names = []
+                query.default_cols = False
+                for f in self._fields:
+                    # we inspect the full extra_select list since we might
+                    # be adding back an extra select item that we hadn't
+                    # had selected previously.
+                    if f in query.extra_select:
+                        extra_names.append(f)
+                    elif f in query.annotation_select:
+                        annotation_names.append(f)
+                    else:
+                        field_names.append(f)
+        else:
+            # Default to all fields.
+            field_names = [f.attname for f in self.model._meta.concrete_fields]
+            extra_names = None
+            annotation_names = None
+
+        query.select = []
+        query.add_fields(field_names, True)
+        if extra_names is not None:
```

Comment:
I wonder why this is behind a `if` block, that means that `[]` can go through (effectively clearing) the mask whereas `None` can't.